### PR TITLE
Embed images in markdown export

### PR DESCRIPTION
**Pull request by xhudaman**
Adds ability to embed images in markdown files.

The way this was achieved is as folllows:
- When exporting markdown it walks the tree of content for image entries
- uses the built in util in the renderer to get the URL for the image asset
- uses `fetch` to get the file contents and convert it to a blob
- uses the `FileReader` API to asynchronously generate a base64 dataURI from the blob

I was thinking it over and it might be a better UX to have the ability to opt into embedding the images as opposed to automatically doing so.  The way I can see this working is expanding the export menu to include new options `Download Chapter as markdown (embeded images)` and `Download Book as markdown (embeded images)` or something similar.  This would allow users who don't want the images to be able to exclude them by just not downloading them from the `5etools-img` repo.  What your thoughts are on it?

### CODE REVIEW

**Comment by revilowaldow** [comment]
Please can this be an option when exporting and not the default. I really don't want 1-2Mb markdown files to balloon to multiple hundreds of Mb when we could just have  instead.

**Comment by xhudaman** [comment]
Gladly!  I'll get the change in today after work.

Do you have any preferred wording for the menu options?  My suggestions were all I could come up with at 3am, but I think they're a bit wordy.

**Comment by xhudaman** [comment]
Might be nice to make that a setting and then use that.

If we fail to load an image we default to `` going forward.  If I wrote the code correctly, that is.  I'll have to confirm that later too.

File: js/bookutils.js [hunk 1]
```diff
@@ -928,6 +936,66 @@ class BookUtil {
...
+			} else {
+				if (input.entries) {
+					await this.walkForImages(input.entries);
+				}
+
+				if (input.type === "image" && !input.data?.base64) {
+					const imageUrl = Renderer.utils.getEntryMediaUrl(entry, "href", "img");
+					(entry.data ||= {}).base64 = await this._getImageDataURL(imageUrl);
+				}
+			}
```
Comment (TheGiddyLimit) [inline js/bookutils.js:968]: there's a generic async walker you can use:

```js
await MiscUtil.getAsyncWalker({keyBlocklist: MiscUtil.GENERIC_WALKER_ENTRIES_KEY_BLOCKLIST, isNoModification: true})
	.pWalk(
		input,
		{
			object: async entry => {
				if (entry.type !== "image" || entry.data?.base64) return;

				const imageUrl = Renderer.utils.getEntryMediaUrl(entry, "href", "img");
				(entry.data ||= {}).base64 = await this._getImageDataURL(imageUrl);
			},
		},
	);
```

though iirc nothing uses it currently, so it might be in need of some TLC

File: js/bookutils.js [hunk 2]
```diff
@@ -928,6 +936,66 @@ class BookUtil {
 	}
 
 	static _getHrefShowAll (bookId) { return `#${UrlUtil.encodeForHash(bookId)},-1`; }
+
+	static async walkForImages (input) {
```
Comment (TheGiddyLimit) [inline js/bookutils.js:940]: `walkForImages` -> `_walkForImages`

File: js/bookutils.js [hunk 3]
```diff
@@ -928,6 +936,66 @@ class BookUtil {
...
+					await this.walkForImages(input.entries);
+				}
+
+				if (input.type === "image" && !input.data?.base64) {
+					const imageUrl = Renderer.utils.getEntryMediaUrl(entry, "href", "img");
+					(entry.data ||= {}).base64 = await this._getImageDataURL(imageUrl);
+				}
+			}
+		} catch (error) {
+			JqueryUtil.doToast({
```
Comment (TheGiddyLimit) [inline js/bookutils.js:970]: for user-facing functionality, we prefer "best effort"/"warn, but do _something_", so here we should:

- default to the "url" version if we can't pull the base64 image
- keep an array of failed URLs (all your errors below could really be "Couldn't load image ${url}", as far as the user's concerned, and more specific logging doesn't help devs in any meaningful way); if there's anything in this array at the end of the loading process, toast the user with a big list of them all 
- log out any real exceptions to the console in the background (using `setTimeout(() => { throw ... })`)

this would also mean wrapping the `_getImageDataURL` internals in a `try`/`catch`, rather than the walker

**Review by TheGiddyLimit — CHANGES_REQUESTED** [review]
all looks fairly reasonable so far

re: list of download options, there _are_ actually some secret global Markdown settings accessible in various "Printer View"s:

image

but it'd all need a proper refactor before we'd want to use it here. I'm happy with

```
Download Chapter as Markdown (With Images)
Download <Adventure|Book> as Markdown (With Images)
```

for now. Stick a `null` action between the two groups (the existing options, and these new ones) so it gets a nice horizontal line separating them, too

**Comment by xhudaman** [comment]
I am close to having the requested changes made.

There are a couple things I want to propose/ask about.

1. When the images have already been fetched (think exporting twice, the first being with embedded images), subsequent downloads without images mistakenly include them since the check whether to use them is determined on whether `data.base64` exists.  Should we create some sort of way to opt in/out that isn't tied to the data existing?  I am thinking something as an interim solution until the settings get re-worked.
2. Would it be preferred to nest the different export options so that they live in nested menus or separate the options to include images from the defaults with a null menu action?

File: js/bookutils.js (hunk 3 above)
Comment (xhudaman) [reply js/bookutils.js:987]: I couldn't figure out how to get the toast to render nicely.  I wanted to go for a `ul` because joining the list with `\n` didn't work and it looked pretty bad and unreadable.

Would love suggestions on how to properly format this toast!

File: js/utils.js [hunk 4]
```diff
@@ -902,7 +902,7 @@ globalThis.JqueryUtil = {
 	_WRP_TOAST: null,
 	_ACTIVE_TOAST: [],
 	/**
-	 * @param {{content: jQuery|string, type?: string, autoHideTime?: boolean} | string} options The options for the toast.
+	 * @param {{content: jQuery|string, type?: string, autoHideTime?: number, isAutoHide?: boolean} | string} options The options for the toast.
```
Comment (xhudaman) [inline js/utils.js:905]: Spotted this was outdated when looking at it for reference, so I decided to fix it.
